src/inmanta/server/services/orchestrationservice.py

```diff
@@ -63,277 +63,302 @@
+        """
+        self.resource_id1 = resource_id1
+        self.resource_id2 = resource_id2
+        super().__init__(self.get_error_message())
+
+    def get_error_message(self) -> str:
+        return (
+            f"A dependency exists between resources {self.resource_id1} and {self.resource_id2}, but they belong to"
+            f" different resource sets."
+        )
+
 
-    def get_resource_id_str(self) -> ResourceIdStr:
-        return self.resource.get_resource_id_str()
+class ResourceSetValidator:
+    def __init__(self, resources: abc.Set[data.Resource]) -> None:
+        self.resources = resources
+        self.rid_to_resource_set = {res.resource_id: res.resource_set for res in self.resources}
 
-    def get_resource_state_for_new_resource(self) -> Literal[ResourceState.undefined, ResourceState.available]:
+    def _is_cross_resource_set_dependency(self, res: data.Resource, rid_dependency: ResourceIdStr) -> bool:
         """
-        Return the resource state as expected by the `resource_state` argument of the `OrchestrationService._put_version()`
-        method. This method should set the resource state of a resource to undefined when it directly depends on an unknown or
-        available when it doesn't.
+        Return True iff the dependency between resource res and the resource with id rid_dependency is a cross-resource set
+        dependency.
         """
-        if self.resource_state is ResourceState.undefined:
-            return ResourceState.undefined
-        else:
-            return ResourceState.available
-
-    def is_shared_resource(self) -> bool:
-        return self.resource_set is None
+        if res.resource_set is None:
+            # Resource is in shared resource set.
+            return False
+        if rid_dependency not in self.rid_to_resource_set:
+            # A partial compile was done and we have a dependency on a resource in another resource set
+            # that is not part of the partial compile.
+            return True
+        resource_set_dep = self.rid_to_resource_set[rid_dependency]
+        if resource_set_dep is None:
+            # Dependency towards shared resource set.
+            return False
+        return res.resource_set != resource_set_dep
 
-    def is_update(self, other: "ResourceWithResourceSet") -> bool:
+    def ensure_no_cross_resource_set_dependencies(self) -> None:
+        """
+        This method raises a CrossResourceSetDependencyError when a resource in self.resources that belongs to a non-shared
+        resource set has a dependency (requires/provides) on another resource that belongs to a different non-shared resource
+        set.
+        """
+        for res in self.resources:
+            # It's sufficient to only check the requires relationship. The provides
+            # relationship is always a subset of the reverse relationship (provides relationship).
+            # The provides relationship only contains the cross-agent dependencies.
```

Comment:
```suggestion
            # It's sufficient to only check the requires relationship. The provides
            # field is always a subset of the reverse relationship (provides relationship).
            # The provides field only contains the cross-agent dependencies.
```
Perhaps this makes it even more clear?